File: services/customer/src/main/java/com/bastion/customer/service/CustomerService.java
package com.bastion.customer.service;

import java.util.UUID;

import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import com.bastion.customer.dto.CustomerResponse;
import com.bastion.customer.dto.UpdateCustomerRequest;
import com.bastion.customer.entity.Customer;
import com.bastion.customer.repository.CustomerRepository;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

@Slf4j
@Service
@RequiredArgsConstructor
public class CustomerService {

    private final CustomerRepository customerRepository;
    // Retrieve customer profile by identity user ID

    @Transactional(readOnly = true)
    public CustomerResponse getProfile(UUID identityUserId) {
        Customer customer = customerRepository.findByIdentityUserId(identityUserId)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Customer profile not found"));
        return CustomerResponse.fromEntity(customer);
    }

    // Update customer profile details (PATCH)
    @Transactional
    public CustomerResponse updateProfile(UUID identityUserId, UpdateCustomerRequest request) {
        Customer customer = customerRepository.findByIdentityUserId(identityUserId)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Customer profile not found"));
        if (request.getFullName() != null) {
            customer.setFullName(request.getFullName().trim());
        }
        if (request.getPhoneNumber() != null) {
            customer.setPhoneNumber(request.getPhoneNumber().trim());
        }
        Customer updatedCustomer = customerRepository.save(customer);
        log.info("Customer profile updated for identityUserId: {}", identityUserId);
        return CustomerResponse.fromEntity(updatedCustomer);
    }

    // Create a new customer from UserRegistered event (Idempotent)
    @Transactional
    public Customer createCustomer(UUID identityUserId, String email, String fullName) {
        // Idempotency check: if customer already exists, skip creation
        if (customerRepository.existsByIdentityUserId(identityUserId)) {
            log.warn("Customer already exists for identityUserId: {}, skipping creation", identityUserId);
            return customerRepository.findByIdentityUserId(identityUserId).orElseThrow();
        }
        Customer newCustomer = Customer.builder()
                .identityUserId(identityUserId)
                .email(email)
                .fullName(fullName)
                .status("ACTIVE")
                .build();
                
        Customer saved = customerRepository.save(newCustomer);
        log.info("New customer created successfully with ID: {} for identityUserId: {}", saved.getId(), identityUserId);
        return saved;
    }
}

File: services/customer/service/customer_service.go
package service

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"strings"
	"time"

	"github.com/agamlatiff/bastion/services/customer/domain"
	"github.com/google/uuid"
	"github.com/redis/go-redis/v9"
)

const (
	profileCachePrefix = "customer:profile:"
	profileCacheTTL    = 30 * time.Minute
)

// CustomerService orchestrates customer profile use cases and Redis cache-aside operations.
type CustomerService struct {
	customerRepo domain.CustomerRepository
	rdb          *redis.Client
}

// NewCustomerService instantiates a new CustomerService.
func NewCustomerService(repo domain.CustomerRepository, rdb *redis.Client) *CustomerService {
	return &CustomerService{
		customerRepo: repo,
		rdb:          rdb,
	}
}

// GetProfile retrieves the customer profile by identity user ID with Redis cache-aside.
func (s *CustomerService) GetProfile(ctx context.Context, identityUserID uuid.UUID) (*domain.CustomerResponse, error) {
	cacheKey := fmt.Sprintf("%s%s", profileCachePrefix, identityUserID.String())

	// 1. Check Redis Cache
	if s.rdb != nil {
		cachedData, err := s.rdb.Get(ctx, cacheKey).Bytes()
		if err == nil {
			var resp domain.CustomerResponse
			if jsonErr := json.Unmarshal(cachedData, &resp); jsonErr == nil {
				return &resp, nil
			}
		}
	}

	// 2. Query Database on Cache Miss
	customer, err := s.customerRepo.FindByIdentityUserID(ctx, identityUserID)
	if err != nil {
		return nil, err
	}

	resp := customer.ToResponse()

	// 3. Populate Redis Cache
	if s.rdb != nil {
		if encoded, err := json.Marshal(resp); err == nil {
			_ = s.rdb.Set(ctx, cacheKey, encoded, profileCacheTTL).Err()
		}
	}

	return &resp, nil
}

// UpdateProfile updates customer details and invalidates the cached profile in Redis.
func (s *CustomerService) UpdateProfile(ctx context.Context, identityUserID uuid.UUID, req domain.UpdateCustomerRequest) (*domain.CustomerResponse, error) {
	customer, err := s.customerRepo.FindByIdentityUserID(ctx, identityUserID)
	if err != nil {
		return nil, err
	}

	if req.FullName != nil {
		trimmed := strings.TrimSpace(*req.FullName)
		customer.FullName = &trimmed
	}
	if req.PhoneNumber != nil {
		trimmed := strings.TrimSpace(*req.PhoneNumber)
		customer.PhoneNumber = &trimmed
	}

	if err := s.customerRepo.Update(ctx, customer); err != nil {
		return nil, err
	}

	// Invalidate Redis cache
	if s.rdb != nil {
		cacheKey := fmt.Sprintf("%s%s", profileCachePrefix, identityUserID.String())
		_ = s.rdb.Del(ctx, cacheKey).Err()
	}

	log.Printf("[CustomerService] Profile updated for identityUserId: %s", identityUserID)
	resp := customer.ToResponse()
	return &resp, nil
}

// CreateCustomer creates a customer profile idempotently upon receiving UserRegistered event.
func (s *CustomerService) CreateCustomer(ctx context.Context, identityUserID uuid.UUID, email string, fullName *string) (*domain.Customer, error) {
	exists, err := s.customerRepo.ExistsByIdentityUserID(ctx, identityUserID)
	if err != nil {
		return nil, err
	}
	if exists {
		log.Printf("[CustomerService] Customer already exists for identityUserId: %s, skipping creation", identityUserID)
		return s.customerRepo.FindByIdentityUserID(ctx, identityUserID)
	}

	var trimmedName *string
	if fullName != nil {
		t := strings.TrimSpace(*fullName)
		if t != "" {
			trimmedName = &t
		}
	}

	newCustomer := &domain.Customer{
		ID:             uuid.New(),
		IdentityUserID: identityUserID,
		Email:          strings.TrimSpace(email),
		FullName:       trimmedName,
		Status:         "ACTIVE",
		CreatedAt:      time.Now().UTC(),
		UpdatedAt:      time.Now().UTC(),
	}

	if err := s.customerRepo.Create(ctx, newCustomer); err != nil {
		return nil, err
	}

	log.Printf("[CustomerService] New customer created successfully with ID: %s for identityUserId: %s", newCustomer.ID, identityUserID)
	return newCustomer, nil
}
